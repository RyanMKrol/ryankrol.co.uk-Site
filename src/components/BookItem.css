.BookItem {
  padding: 15px;
  width: 220px;
  text-align: left;
  background-color: #eee;
  margin: 10px;
  padding-bottom: 0;
  font-family: 'Archivo Narrow', sans-serif;
  font-size: 14px;
  border-radius: 5px;
  display: flex;
  flex-direction: column;
}

.book-item-thumbnail {
  text-align: center;
  align-items: center;
}

.book-item-image {
  max-width: 100%;
  height: 200px;
  display: flex;
  margin: auto;
  border-radius: 5px;
  border-style: solid;
  border-color: #FDD023;
  border-width: 8px;
}

.book-information {
  display: flex;
  flex: 1;
}

.book-information-content {
  margin: auto 0;
  padding: 16px 0;
}

.book-information-key {
  color: #555;
}

.book-information-value {
  color: #888;
}
