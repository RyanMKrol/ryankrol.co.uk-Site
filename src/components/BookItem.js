import React, { Component } from 'react';

import './BookItem.css';

/**
 * Object representing the images returned for a book
 *
 * @typedef {object} BookItemImages
 * @property {string} thumbnail Image used for the thumbnail of the book
 */

/**
 * Object representing the props to an AlbumItem
 *
 * @typedef {object} BookItemProps
 * @property {string} title Title of the book
 * @property {Array<string>} authors Authors of the book
 * @property {BookItemImages} images Images item used the thumbnail
 * @property {string} isbn ISBN for the book
 * @property {string} link Link to buy the book
 */

/**
 * Class representing a UI component for a book
 */
class BookItem extends Component {
  /**
   * @param {BookItemProps} props Arguments to build the widget
   */
  constructor(props) {
    super();

    this.title = props.title;
    this.authors = props.authors.join(', ');
    this.images = props.images;
    this.isbn = props.isbn;
    this.link = `https://www.amazon.co.uk/s?k=${this.isbn}`;

    if (!(this.title && this.authors && this.images && this.isbn)) {
      throw new Error('Did not pass all required args for rendering BookItem');
    }
  }

  /**
   * Render method
   *
   * @returns {string} The JSX representing the component
   */
  render() {
    return (
      <div className="BookItem">
        <div className="book-item-thumbnail">
          <img
            className="book-item-image"
            src={this.images.thumbnail}
            alt={`${this.title} - ${this.authors}`}
          />
        </div>
        <div className="book-information">
          <p className="book-information-content">
            <span className="book-information-key">Title - </span>
            <span className="book-information-value">{this.title}</span>
            <br />
            <span className="book-information-key">
              {this.props.authors.length > 1 ? 'Authors' : 'Author'} -{' '}
            </span>
            <span className="book-information-value">{this.authors}</span>
            <br />
            <span className="book-information-key">ISBN - </span>
            <span className="book-information-value">
              <a href={this.link}>{this.isbn}</a>
            </span>
          </p>
        </div>
      </div>
    );
  }
}

export default BookItem;
