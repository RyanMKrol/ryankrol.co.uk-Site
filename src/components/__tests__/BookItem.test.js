import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import BookItem from './../BookItem'

const defaultArgs = {
  title: 'test-title',
  authors: ['test-authors'],
  images: 'test-images',
  isbn: 'test-isbn',
  bookId: 'test-bookId',
}

// Render Tests

it('renders BookItem correctly', () => {
    const bookItem = shallow(
      <BookItem {...defaultArgs} />)
    expect(toJson(bookItem)).toMatchSnapshot()
})

it('throws when required parameters are omitted', () => {
    expect(() => mount(<BookItem />)).toThrow()
})

it('does not throw when required parameters are provided', () => {
    expect(() => mount(<BookItem {...defaultArgs}/>)).not.toThrow()
})
