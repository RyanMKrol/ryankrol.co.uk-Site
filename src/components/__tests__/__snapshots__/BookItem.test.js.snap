// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders BookItem correctly 1`] = `
<div
  className="BookItem"
>
  <link
    href="https://fonts.googleapis.com/css?family=Archivo+Narrow&display=swap"
    rel="stylesheet"
  />
  <div
    className="book-item-thumbnail"
  >
    <img
      alt="test-title - test-authors"
      className="book-item-image"
    />
  </div>
  <div
    className="book-information"
  >
    <p
      className="book-information-content"
    >
      <span
        className="book-information-key"
      >
        Title - 
      </span>
      <span
        className="book-information-value"
      >
        test-title
      </span>
      <br />
      <span
        className="book-information-key"
      >
        Authors - 
      </span>
      <span
        className="book-information-value"
      >
        test-authors
      </span>
      <br />
      <span
        className="book-information-key"
      >
        ISBN - 
      </span>
      <span
        className="book-information-value"
      >
        test-isbn
      </span>
    </p>
  </div>
</div>
`;
